refactor: delete all trackers when finalizing CCTX

kingpinXD opened the pull request:
# Description

Closes https://github.com/zeta-chain/node/issues/2610
- Try to delete all trackers associated with the cctx when finalizing it. If the tracker doesn't exist, nothing happens
- Try to update the pending nonce for all nonces associated with the tracker, if the nonce has been previously updated, nothing happens 
- Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )

# How Has This Been Tested?

<!--- Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. Link any optional github actions runs. -->

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Added a validation check for outbound transaction trackers in the local end-to-end testing framework.
  - Introduced a new method to set the outbound ballot index within the CrossChainTx structure.
  
- **Bug Fixes**
  - Improved handling of outbound transaction states by simplifying method signatures and enhancing functionality.

- **Documentation**
  - Minor formatting adjustments in command-line documentation for clarity.

- **Tests**
  - Expanded test coverage for outbound transaction handling and nonce removal functionality.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

kingpinXD commented on x/crosschain/types/cctx_test.go:
```diff
@@ -16,6 +16,12 @@ import (
 	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
 )
 
+func TestCrossChainTx_SetOutboundBallot(t *testing.T) {
+	cctx := sample.CrossChainTx(t, "test")
+	ballotIndex := sample.ZetaIndex(t)
+	cctx.SetOutboundBallotIndex(ballotIndex)
+	require.Equal(t, ballotIndex, cctx.GetCurrentOutboundParam().BallotIndex)
+}
```
That's a good suggestion

kingpinXD commented on cmd/zetae2e/local/local.go:
```diff
@@ -362,10 +362,18 @@ func localE2ETest(cmd *cobra.Command, _ []string) {
 	os.Exit(0)
 }
 
+func checkTrackers(ctx context.Context, deployRunner *runner.E2ERunner) {
+	// get all trackers
+	res, err := deployRunner.CctxClient.OutTxTrackerAll(ctx, &crosschaintypes.QueryAllOutboundTrackerRequest{})
+	require.NoError(deployRunner, err)
+	require.Empty(deployRunner, res.OutboundTracker, "there should be no trackers")
+}
```
This check is not needed as this is test, and this call should not return an error

kingpinXD commented on x/crosschain/keeper/msg_server_vote_outbound_tx_test.go:
```diff
@@ -465,7 +469,7 @@ func TestKeeper_SaveFailedOutbound(t *testing.T) {
 		})
 		cctx.CctxStatus.Status = types.CctxStatus_PendingOutbound
-		k.SaveFailedOutbound(ctx, cctx, sample.String(), sample.ZetaIndex(t))
+		k.SaveFailedOutbound(ctx, cctx, sample.String())
```
This suggestion is wrong , the string argument is required for the error message

lumtis commented on cmd/zetae2e/local/local.go:
```diff
@@ -366,6 +366,16 @@ func localE2ETest(cmd *cobra.Command, _ []string) {
 	os.Exit(0)
 }
 
+func checkTrackers(deployRunner *runner.E2ERunner) {
```
Could be implemented in `e2e` package instead

lumtis commented on x/observer/keeper/pending_nonces_test.go:
```diff
@@ -136,4 +137,54 @@ func TestKeeper_RemoveFromPendingNonces(t *testing.T) {
+		tss := sample.Tss()
+		// make nonces and pubkey deterministic for test
+		chainIDS := []int64{chains.GoerliLocalnet.ChainId, chains.BitcoinTestnet.ChainId, chains.BscTestnet.ChainId}
+		pn := make([]types.PendingNonces, len(chainIDS))
```
```suggestion
		pendingNonces := make([]types.PendingNonces, len(chainIDS))
```
better to keep explicit variable names

lumtis commented on x/observer/keeper/pending_nonces_test.go:
```diff
@@ -136,4 +137,54 @@ func TestKeeper_RemoveFromPendingNonces(t *testing.T) {
+
+		// remove from pending nonces
+		k.RemoveFromPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId, 1)
+		pnGoerli, found := k.GetPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId)
```
```suggestion
		actualPendingNonces, found := k.GetPendingNonces(ctx, tss.TssPubkey, chains.GoerliLocalnet.ChainId)
```

lumtis commented on x/crosschain/types/cctx.go:
```diff
@@ -164,6 +164,10 @@ func (m CrossChainTx) GetCCTXIndexBytes() ([32]byte, error) {
 	return sendHash, nil
 }
 
+func (m CrossChainTx) SetOutboundBallotIndex(index string) {
```
Let's add comment, it adds the index specifically to the last outbound

lumtis reviewed (commented):
>Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )

What is the change reflecting this?

kingpinXD commented:
> > Fixes setting the ballot index properly for cases when a revert is added . (previously the ballot index was getting set to the revert tx instead of the original outbound )
> 
> What is the change reflecting this?

https://github.com/zeta-chain/zeta-node/blob/22f14acd3f5e32aeb370866cb6f3ab2576865fe5/x/crosschain/keeper/msg_server_vote_outbound_tx.go#L93-L94

Earlier this ballot was set in the SaveOutbound. 
In case if a revert was added , the ballot index for outboundParam[0] was in fact added to outboundParam[1]

lumtis reviewed (approved):
LGTM
@kingpinXD please use explicit PR title

kingpinXD commented:
> LGTM @kingpinXD please use explicit PR title

@lumtis could you give me an example of what you mean, Imo the title reflects the change 
Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )

lumtis commented:
> > LGTM @kingpinXD please use explicit PR title
> 
> @lumtis could you give me an example of what you mean, Imo the title reflects the change Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )

I changed the name from `tracker cleanup` to `delete all trackers when finalizing CCTX` before putting this comment.
 `tracker cleanup` doesn't allow to understand the change by looking at PR title.

kingpinXD commented:
> > > LGTM @kingpinXD please use explicit PR title
> > 
> > 
> > @lumtis could you give me an example of what you mean, Imo the title reflects the change Are you referring to the fact that pending nonces are also cleaned ? ( I did not add that in as the change keeps the same behaviour as before )
> 
> I changed the name from `tracker cleanup` to `delete all trackers when finalizing CCTX` before putting this comment. `tracker cleanup` doesn't allow to understand the change by looking at PR title.

okay that makes sense , thanks for the edit . 
Got confused that you were requesting a further change
